src/System.Text.Utf8/src/System/Text/Utf8/Utf8StringEnumerator.cs

```diff
@@ -0,0 +1,148 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+
+namespace System.Text.Utf8
+{
+    // TODO: This object is heavier than string itself... Once we got ByteSpan runtime support, change it.
+    public struct Utf8StringEnumerator : IEnumerator<UnicodeCodePoint>
+    {
+        private ByteSpan _startBuffer;
+        private ByteSpan _buffer;
+
+        private byte[] _bytes;
+        private int _startIndex;
+        private int _index;
+        private int _lastIndex;
+
+        private int _currentLenCache;
+
+        public Utf8StringEnumerator(byte[] bytes, int index, int length)
+        {
+            if (index + length > bytes.Length)
+                throw new ArgumentOutOfRangeException("index");
+
+            _startBuffer = default(ByteSpan);
+            _buffer = default(ByteSpan);
+
+            _bytes = bytes;
+            // first MoveNext moves to a first byte
+            _startIndex = index - 1;
+            _index = _startIndex;
+            _lastIndex = index + length;
+
+            _currentLenCache = 0;
+        }
+
+        public unsafe Utf8StringEnumerator(ByteSpan buffer)
+        {
+            if (buffer.UnsafeBuffer == null)
+                throw new ArgumentNullException("buffer");
+
+            // first MoveNext moves to a first byte
+            _startBuffer = new ByteSpan(buffer.UnsafeBuffer - 1, buffer.Length + 1);
+            _buffer = _startBuffer;
+
+            _bytes = default(byte[]);
+            _startIndex = default(int);
+            _index = default(int);
+            _lastIndex = default(int);
+
+            _currentLenCache = 0;
+        }
+
+
+        object IEnumerator.Current { get { return Current; } }
+
+        public unsafe UnicodeCodePoint Current
+        {
+            get
+            {
+                UnicodeCodePoint ret;
+                bool succeeded;
+                if (_bytes != null)
+                {
+                    if (_index == _startIndex)
+                        throw new InvalidOperationException();
+
+                    fixed (byte* pinnedBytes = _bytes)
+                    {
+                        ByteSpan buffer = new ByteSpan(pinnedBytes + _index, _lastIndex - _index);
+
+                        succeeded = Utf8Encoder.TryDecodeCodePoint(buffer, out ret, out _currentLenCache);
+                    }
+                }
+                else
+                {
+                    if (_buffer.UnsafeBuffer == _startBuffer.UnsafeBuffer)
```

Comment:
I agree about the bugs resulting from that but also makes it slightly less readable since the text becomes way vertically wider. I personally like doing that for checking initial conditions and rarely do it in other places. Will change it